homeassistant/components/recorder/__init__.py

```diff
@@ -314,127 +352,198 @@ def do_adhoc_purge(self, **kwargs):
+    @callback
+    def async_register(self, shutdown_task, hass_started):
+        """Post connection initialize."""
 
-        if not self._setup_recorder():
-            return
+        def shutdown(event):
+            """Shut down the Recorder."""
+            if not hass_started.done():
+                hass_started.set_result(shutdown_task)
+            self.queue.put(None)
+            self.join()
 
-        shutdown_task = object()
-        hass_started = concurrent.futures.Future()
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, shutdown)
+
+        if self.hass.state == CoreState.running:
+            hass_started.set_result(None)
+            return
 
         @callback
-        def register():
-            """Post connection initialize."""
-            self.async_db_ready.set_result(True)
+        def async_hass_started(event):
+            """Notify that hass has started."""
+            hass_started.set_result(None)
 
-            def shutdown(event):
-                """Shut down the Recorder."""
-                if not hass_started.done():
-                    hass_started.set_result(shutdown_task)
-                self.queue.put(None)
-                self.join()
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STARTED, async_hass_started)
 
-            self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, shutdown)
+    @callback
+    def async_connection_failed(self):
+        """Connect failed tasks."""
+        self.async_db_ready.set_result(False)
+        persistent_notification.async_create(
+            self.hass,
+            "The recorder could not start, check [the logs](/config/logs)",
+            "Recorder",
+        )
+        self._stop_queue_watcher_and_event_listener()
 
-            if self.hass.state == CoreState.running:
-                hass_started.set_result(None)
-            else:
+    @callback
+    def async_connection_success(self):
+        """Connect success tasks."""
+        self.async_db_ready.set_result(True)
 
-                @callback
-                def notify_hass_started(event):
-                    """Notify that hass has started."""
-                    hass_started.set_result(None)
+    @callback
+    def _async_recorder_ready(self):
+        """Mark recorder ready."""
+        self.async_recorder_ready.set()
 
-                self.hass.bus.async_listen_once(
-                    EVENT_HOMEASSISTANT_START, notify_hass_started
-                )
+    @callback
+    def async_purge(self, now):
+        """Trigger the purge."""
+        self.queue.put(PurgeTask(self.keep_days, repack=False, apply_filter=False))
 
-        self.hass.add_job(register)
-        result = hass_started.result()
+    def run(self):
+        """Start processing events to save."""
+        shutdown_task = object()
+        hass_started = concurrent.futures.Future()
+
+        self.hass.add_job(
+            callback(lambda: self.async_register(shutdown_task, hass_started))
```

Comment:
This should work too (with `from functools import partial`)

```suggestion
            partial(self.async_register, shutdown_task, hass_started)
```